### PR TITLE
Path is empty for VB My namespace

### LINKED ISSUE #775: The path is empty. (Parameter 'path')
Having an issue generating the report. The project is SDK format but references older projects. When I run'

(actual paths omitted)
`dotnet test "....csproj" --configuration Release --filter Category!=IntegrationTest --logger:trx --results-directory "...\TestResults" --settings "....runsettings" /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura /p:CoverletOutput="....Tests.cobertura.xml"`

I get the following error

> Calculating coverage result...
>   Generating report '....Tests.cobertura.xml'
> ...\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error : The path is empty. (Parameter 'path') [....Tests.csproj]
> ...nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at System.IO.Path.GetFullPath(String path) [...Tests.csproj]
> ...\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at System.IO.Directory.GetDirectoryRoot(String path) [....Tests.csproj]
> ....nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at System.Linq.Lookup`2.Create(IEnumerable`1 source, Func`2 keySelector, IEqualityComparer`1 comparer) [....Tests.csproj]
> ..\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at System.Linq.GroupedEnumerable`2.GetEnumerator() [...Tests.csproj]
> ...\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at System.Linq.Enumerable.SelectEnumerableIterator`2.ToList() [...Tests.csproj]
> ...\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.Core.Reporters.CoberturaReporter.Report(CoverageResult result) in D:\git\coverletToRelease\src\coverlet.core\Reporters\CoberturaReporter.cs:line 34 [...Tests.csproj]
> C:\Users\aharris\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.MSbuild.Tasks.ReportWriter.WriteReport() in D:\git\coverletToRelease\src\coverlet.msbuild.tasks\ReportWriter.cs:line 50 [....Tests.csproj]
> ...\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(36,5): error :    at Coverlet.MSbuild.Tasks.CoverageResultTask.Execute() in D:\git\coverletToRelease\src\coverlet.msbuild.tasks\CoverageResultTask.cs:line 145 [...Tests.csproj]

This works for several other projects in the solution that don't have a reference to older projects. 

I also noticed I get these warnings when building:

> ...\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(7,5): warning : [coverlet] Unable to instrument module: ...Tests\bin\Debug\net472\Microsoft.ReportViewer.Common.dll because : No symbol found for file: ....Tests\bin\Debug\net472\Microsoft.ReportViewer.Common.dll [....Tests.csproj]
> ...\aharris\.nuget\packages\coverlet.msbuild\2.8.0\build\coverlet.msbuild.targets(7,5): warning : [coverlet] Unable to instrument module: ....Tests\bin\Debug\net472\Microsoft.ReportViewer.WebForms.dll because : No symbol found for file: ....Tests\bin\Debug\net472\Microsoft.ReportViewer.WebForms.dll [....Tests.csproj]

### PR DESCRIPTION
**Pull request by daveMueller**
closes #775

### CODE REVIEW

**Comment by daveMueller** [comment]
The issue is the ```My``` namespace in Visual Basic ([My](https://docs.microsoft.com/en-us/dotnet/visual-basic/developing-apps/development-with-my/)). Usually a sequence point has a reference to a source file like this.

grafik

In Visual Basic the compiler injects the ```My``` classes into the assembly without a reference to a source file.

grafik

Maybe there are also other scenarios where a sequence point doesn't have a reference to a source file that we haven't encountered yet. So I created a PR that excludes code like this within an assembly.

But I could also think of a solution that really just filters out the ```My``` namespace in a Visual Basic assembly. Just let me know what you think.

**Comment by MarcoRossignoli** [comment]
Thx Dave, need to go deeper on it to understand if as you said make sense something for that injected namespace.

**Comment by daveMueller** [comment]
> Can we add a new VB project and create a test?Is it too complex?
> Something like `coverlet.tests.projectsample.vbmynamespace` here https://github.com/coverlet-coverage/coverlet/tree/master/test

I will give it a try. 
But nevertheless I think instrumentation of IL code that doesn't have a reference to a source file doesn't make much sense for the coverage.

**Comment by MarcoRossignoli** [comment]
> But nevertheless I think instrumentation of IL code that doesn't have a reference to a source file doesn't make much sense for the coverage.

I agree...but we need to be pretty sure to not break something, so I'm ok with the update but I'd like to have a test in place for future issue on it.

File: src/coverlet.core/Instrumentation/Instrumenter.cs [hunk 1]
```diff
@@ -511,6 +511,9 @@ private void InstrumentType(TypeDefinition type)
         private void InstrumentMethod(MethodDefinition method)
         {
             var sourceFile = method.DebugInformation.SequencePoints.Select(s => _sourceRootTranslator.ResolveFilePath(s.Document.Url)).FirstOrDefault();
+
+            if (string.IsNullOrEmpty(sourceFile)) return;
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Instrumentation/Instrumenter.cs:515]: Can you add an information log with method/module information

File: test/coverlet.core.tests/Instrumentation/InstrumenterTests.cs [hunk 2]
```diff
@@ -597,7 +597,7 @@ public int SampleMethod()
             }
             else
             {
-                Assert.NotEmpty(result.Documents);
+                Assert.Empty(result.Documents);
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Instrumentation/InstrumenterTests.cs:600]: mmm is this correct?

Comment (daveMueller) [reply test/coverlet.core.tests/Instrumentation/InstrumenterTests.cs:600]: The compiled assembly in this test doesn't have a reference to a source file an thus the result is always empty. Before my change in the ```Instrumenter``` this test returned a result with a empty document source which is exactly what would create the bug in the report generation.

image

After my change such a ```InstrumenterResult``` isn't created anymore and  that's why now ```result.Documents``` is also empty here.

To make this test the way it was intended the dynamically created assembly must contain sequence points with a reference to the source file. But what ever I tried here the assemblies sequence points never had source file information.

https://github.com/coverlet-coverage/coverlet/blob/4bbcd0bde26939da588544966091485dfa48a038/test/coverlet.core.tests/Instrumentation/InstrumenterTests.cs#L556

The assemblies sequence points were always without source information.

image

**Comment by daveMueller** [comment]
@MarcoRossignoli It seems the My namespace in VB is only in .NET Framework projects. I think it isn't possible to add a project with .NET Framework as target to the coverlet solution? Or at least I couldn't do it?

File: src/coverlet.core/Instrumentation/Instrumenter.cs (hunk 1 above)
Comment (daveMueller) [reply src/coverlet.core/Instrumentation/Instrumenter.cs:517]: Done

**Comment by MarcoRossignoli** [comment]
> @MarcoRossignoli It seems the My namespace in VB is only in .NET Framework projects. I think it isn't possible to add a project with .NET Framework as target to the coverlet solution? Or at least I couldn't do it without crashing the build jobs?

I think it's possible, AFAIK .NET Framework project can be compiled on all plat through `dotnet` driver, I have some doubt on type resolution for Mac/Linux but in those case we can skip tests. Give it a try and let me know.

**Comment by daveMueller** [comment]
@MarcoRossignoli I'm still working on this but it seems that Linux and Mac build jobs can't build the project. Windows build job is green now. But for the others I'm getting (```The reference assemblies for .NETFramework,Version=v4.7.2 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application.```)

**Comment by MarcoRossignoli** [comment]
Try with this https://andrewlock.net/using-reference-assemblies-to-build-net-framework-libararies-on-linux-without-mono/

https://www.nuget.org/packages/Microsoft.NETFramework.ReferenceAssemblies/

**Comment by daveMueller** [comment]
Thanks, but now I  get a ```vbc : error BC2017: could not find library 'Microsoft.VisualBasic.dll' ``` which is weird because the ReferenceAssemblies package contains the dll. Somehow the build doesn't look in the right place? According to the documentation ([docs](https://github.com/microsoft/dotnet/tree/master/releases/reference-assemblies#classic-style-project)) this should work. Any other ideas? Otherwise  I'll open a issue in the dotnet repo.

**Comment by MarcoRossignoli** [comment]
Ok Dave open an issue on the dotnet repo, I never used that ref I only know about the existence

File: test/coverlet.core.tests/Instrumentation/InstrumenterTests.cs (hunk 2 above)
Comment (MarcoRossignoli) [reply test/coverlet.core.tests/Instrumentation/InstrumenterTests.cs:600]: got it...pls can you add a comment on top of that assert with this explanation for future reference?

**Comment by nk111** [comment]
Hi, I've been trying to get a coverage report for a .Net Framework project displayed in Devops for 3 days. After updating all possible tools on our build agent, the test is finally running via DotNetCli. But unfortunately I also get the "The path is empty" error at the end. It is a VB project.
Can this PR please be closed?

**Comment by MarcoRossignoli** [comment]
@daveMueller do you plan to move on this one or can I take over? It's pretty old and I think we can exclude tests from Mac/Linux and run failing one only on Win.

File: src/coverlet.core/Instrumentation/Instrumenter.cs [hunk 3]
```diff
@@ -510,7 +510,12 @@ private void InstrumentType(TypeDefinition type)
 
         private void InstrumentMethod(MethodDefinition method)
         {
+            //_logger.LogInformation($"Instrumenting method '{method.FullName}' in module '{_module}'.");
```
[the remark this answers is not in the document]
Comment (daveMueller) [reply src/coverlet.core/Instrumentation/Instrumenter.cs:513]: I'm pretty sure I just used it for debugging. So no need to keep this in the log.

**Comment by MarcoRossignoli** [comment]
Great, thanks a lot for the effort here!

**Comment by MarcoRossignoli** [comment]
Is this good to merge now?

**Comment by daveMueller** [comment]
@MarcoRossignoli yes I worked on it again and made the integration test run on all platforms now. So this is ready for review/merge. I asked in the issue (https://github.com/coverlet-coverage/coverlet/issues/775#issuecomment-1401168769) if someone can give it a try.

**Review by MarcoRossignoli — APPROVED** [review]

**Comment by MarcoRossignoli** [comment]
Thanks @daveMueller

**Pull request merged**